lib/player.js

```diff
@@ -4837,20 +4879,42 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
       this.currentTextLanguage_ = stream.language;
     } else if (this.video_ && this.video_.src && this.video_.textTracks) {
       const textTracks = this.getFilteredTextTracks_();
-      for (const textTrack of textTracks) {
-        if (shaka.util.StreamUtils.html5TrackId(textTrack) == track.id) {
-          // Leave the track in 'hidden' if it's selected but not showing.
-          textTrack.mode = this.isTextVisible_ ? 'showing' : 'hidden';
-        } else {
-          // Safari allows multiple text tracks to have mode == 'showing', so be
-          // explicit in resetting the others.
-          textTrack.mode = 'disabled';
+      const oldTrack = textTracks.find((textTrack) =>
+        textTrack.mode !== 'disabled');
+      const newTrack = textTracks.find((textTrack) =>
+        shaka.util.StreamUtils.html5TrackId(textTrack) === track.id);
+      if (oldTrack !== newTrack) {
+        if (oldTrack) {
+          oldTrack.mode = 'disabled';
+          this.loadEventManager_.unlisten(oldTrack, 'cuechange');
+          this.textDisplayer_.remove(0, Infinity);
+        }
+        if (newTrack) {
+          this.enableNativeTrack_(newTrack);
         }
       }
       this.onTextChanged_();
     }
   }
 
+  /**
+   * @param {!TextTrack} track
+   * @private
+   */
+  enableNativeTrack_(track) {
+    this.loadEventManager_.listen(track, 'cuechange', () => {
+      // Always remove cues from the past to avoid memory grow.
+      const removeEnd = Math.max(0,
+          this.video_.currentTime - this.config_.streaming.bufferBehind);
+      this.textDisplayer_.remove(0, removeEnd);
+      const cues = Array.from(track.activeCues || [])
+          .map(shaka.text.Utils.mapNativeCue)
+          .filter(shaka.util.Functional.isNotNull);
+      this.textDisplayer_.append(cues);
+    });
+    track.mode = 'hidden';
```

Comment:
I think you should do something like this:

```
this.loadEventManager_.listen(this.video_, 'enterpictureinpicture', () => {
  track.mode = 'showing';
});
this.loadEventManager_.listen(this.video_, 'leavepictureinpicture', () => {
  track.mode = 'hidden';
});
track.mode = !!document.pictureInPictureElement ? 'showing' : 'hidden';
```

With this I think we would support the PiP with native Safari subtitles.